### PR TITLE
Addresses #3147, add HeatExchanger:Desiccant:BalancedFlow

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

Companion to: https://github.com/NREL/OpenStudio/pull/4317.

Link to relevant GitHub Issue(s) if appropriate: https://github.com/NREL/OpenStudio/issues/3147.

Link to the **Linux.deb** installer to use for CI Testing. If not set, it will default to latest official release.
[OpenStudio Installer]: http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/incremental/develop/4317/OpenStudio-3.2.1-alpha%2Befcf576dba-Ubuntu-18.04.deb

This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,
 - [ ] **Case 2 - `TestFix`:** a fix for an existing test. The GitHub issue should be referenced in the PR description
 - [ ] **Case 3 - `NewTestForExisting`:** a new test for an already-existing model API class
 - [ ] **Case 4 - `Other`:** Something else, like maintenance of the repo, or just committing test results with a new OpenStudio version.

Depending on your answer, please fill out the required section below, and delete the three others.
Leave the review checklist in place.

----------------------------------------------------------------------------------------------------------

### Case 1: New test for a new model API class

Please include which class(es) you are adding a test to specifically test for.
Include a link to the OpenStudio Pull Request in which you are adding the new classes, or the class itself if already on develop.

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [ ] Tests pass either:
     - [ ] with official OpenStudio release (include version):
         - [ ] A matching OSM test has been added from the successful run of the Ruby one with the official OpenStudio release
         - [ ] The label `AddedOSM` has been added to this PR
         - [ ] All new `out.osw` have been committed

     - [ ] with current develop (incude SHA):
         - [ ] The label `PendingOSM` has been added to this PR
         - [ ] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO: To be added in the next official release after: 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [ ] No `out.osw` have been committed as they need to be run with an official OpenStudio version

 - [ ] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
    Please paste the heatmap png generated after running the following commands:
     - [ ] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [ ] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
    Please paste the text output or heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

 - [ ] **Object has been added to `autosize_hvac.rb` to ensure the autosizedXXX values methods do work**

----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [ ] Code style (indentation, variable names, strip trailing spaces)
 - [ ] Functional code review (it has to work!)
 - [ ] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [ ] Appropriate `out.osw` have been committed
 - [ ] Test is stable
 - [ ] Object is tested in `autosize_hvac` as appropriate
 - [ ] The appropriate labels have been added to this PR:
   - [ ] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [ ] If `NewTest`: add `PendingOSM` or `AddedOSM`

### CODE REVIEW

File: model_tests.rb [hunk 1]
```diff
@@ -495,6 +495,15 @@ def test_heatexchanger_airtoair_sensibleandlatent_rb
     result = sim_test('heatexchanger_airtoair_sensibleandlatent.rb')
   end
 
+  def test_heatexchanger_desiccant_balancedflow_rb
+    result = sim_test('heatexchanger_desiccant_balancedflow.rb')
+  end
+
+  # TODO: To be added in the next official release after: 3.1.0
```
Comment (jmarrec) [inline model_tests.rb]:
```suggestion
  # TODO: To be added in the next official release after: 3.2.0
```

File: model/simulationtests/heatexchanger_desiccant_balancedflow.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+model.add_hvac({ 'ashrae_sys_num' => '03' })
+
+air_systems = model.getAirLoopHVACs
+
+air_systems.each_with_index do |s, i|
+  oa_node = s.airLoopHVACOutdoorAirSystem.get.outboardOANode.get
+
+  if i == 0
+    p = OpenStudio::Model::HeatExchangerDesiccantBalancedFlowPerformanceDataType1.new(model)
+    hx = OpenStudio::Model::HeatExchangerDesiccantBalancedFlow.new(model, p)
```
Comment (jmarrec) [inline model/simulationtests/heatexchanger_desiccant_balancedflow.rb]: Please exercise ALL setters for both objects at least once (you can copy all defaults from the model Ctor and just prepend with `hx.` or `.p` as appropriate, that is fine). The goal is both to test the object but also to demonstrate how to use its api. Also, harcoding values in the ruby test means that should we adjust the default in model ctor, we wouldn't see a deviation in EUI suddenly.

File: model/simulationtests/heatexchanger_desiccant_balancedflow.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+model = BaselineModel.new
+
+# make a 2 story, 100m X 50m, 10 zone core/perimeter building
+model.add_geometry({ 'length' => 100,
+                     'width' => 50,
+                     'num_floors' => 1,
+                     'floor_to_floor_height' => 4,
+                     'plenum_height' => 1,
+                     'perimeter_zone_depth' => 3 })
```
Comment (jmarrec) [inline model/simulationtests/heatexchanger_desiccant_balancedflow.rb]: As far as I can tell, you really want to end up with 2 zones, so you have to 2 AirLoopHVAC to connect to. Your comment is outdated in any case. Bottom line, I suggest using this:

```suggestion
# make a 2 story, 100m X 50m, 2 zone core building
model.add_geometry({ 'length' => 100,
                     'width' => 50,
                     'num_floors' => 2,
                     'floor_to_floor_height' => 4,
                     'plenum_height' => 0,
                     'perimeter_zone_depth' => 0 })
```

File: model_tests.rb [hunk 4]
```diff
@@ -495,6 +495,15 @@ def test_heatexchanger_airtoair_sensibleandlatent_rb
     result = sim_test('heatexchanger_airtoair_sensibleandlatent.rb')
   end
 
+  def test_heatexchanger_desiccant_balancedflow_rb
+    result = sim_test('heatexchanger_desiccant_balancedflow.rb')
+  end
```
Comment (jmarrec) [inline model_tests.rb:509]: Another question: I think there's also a relation to CoilSystemCoolingDXHeatExchangerAssisted and CoilSystemCoolingWaterHeatExchangerAssisted. We should test for these too no?

**Review by jmarrec — CHANGES_REQUESTED** [review]

File: model_tests.rb (hunk 4 above)
Comment (joseph-robertson) [reply model_tests.rb:509]: Added two new tests for these.

Comment (jmarrec) [reply model_tests.rb:509]: You also need to add it to autosize_hvac test since the PerfData can be autosized

Comment (joseph-robertson) [reply model_tests.rb:509]: Meaning we don't want to autosize in tests outside of the autosize_hvac test?

Comment (jmarrec) [reply model_tests.rb:509]: not necesarilly:you can use Autosized values in the class test). I mean the autosizing test will pick up that it was not added but can be autosized and will fail. Run it locally and see for yourself (edit: I see you actually already modified it)

The autosize test is responsible for making sure the SQL query to retrieve the autosized value are working correcltly.

 `/path/to/openstudio model_tests.rb -n test_autosizing_rb`

Comment (jmarrec) [reply model_tests.rb:509]: This makes me realize I should modify the pull_request.yml workflow to always run that one...

Comment (jmarrec) [reply model_tests.rb:509]: Added that to the workflow in b84ffa54324fce288e6023219965e8fbbcb8bbb3. I merged develop in, CI should report on it. I also updated the installer link in your OP with the latest form your branch: https://github.com/NREL/OpenStudio/pull/4317#issuecomment-854011082

**Comment by jmarrec** [comment]
There are a few minor errors that I'll correct quickly, but there's one that I found weird.

https://github.com/NREL/OpenStudio-resources/blob/1ae657810848fce06f96b7df320fe2157b34aac5/model/simulationtests/autosize_hvac.rb#L8-L14

So the comment is mine, and it's clearly how I expected things given the signature. the autosize_hvac has actually 40 zones only... Turns out the "plenum_height" parameter is NOT used.

**Comment by jmarrec** [comment]
There's a problem with the CoilSystem:Cooling:Water:HeatExchangerAssisted test... I don't think E+ actually accepts it. I d

```
CoilSystem:Cooling:Water:HeatExchangerAssisted,
  A2 , \field Heat Exchanger Object Type
       \required-field
       \type choice
       \key HeatExchanger:AirToAir:FlatPlate
       \key HeatExchanger:AirToAir:SensibleAndLatent
```

the DX does (but the test fails as nodes aren't written!)
```
CoilSystem:Cooling:DX:HeatExchangerAssisted,
  A2 , \field Heat Exchanger Object Type
       \required-field
       \type choice
       \key HeatExchanger:AirToAir:FlatPlate
       \key HeatExchanger:AirToAir:SensibleAndLatent
       \key HeatExchanger:Desiccant:BalancedFlow
```

**Edit**: confirmed in E+ source code https://github.com/NREL/EnergyPlus/blob/b8f984adb5fe45cfebf838ab1fcf21d104e5091a/src/EnergyPlus/HVACHXAssistedCoolingCoil.cc#L647-L656


(This proves the value of having this suite of integration tests!)

**Review by jmarrec — APPROVED** [review]

**Pull request merged**